### PR TITLE
fix: Change power level without changing memory

### PR DESCRIPTION
**Pull request by krille-chan**
The problem here is that we
have not created a deep copy
of the power level map. By using .copy() we create a
deep copy now.

Closes https://github.com/famedly/product-management/issues/2412

### CODE REVIEW

File: lib/src/room.dart [hunk 1]
```diff
@@ -1187,22 +1187,26 @@ class Room {
   /// Set the power level of the user with the [userID] to the value [power].
   /// Returns the event ID of the new state event. If there is no known
   /// power level event, there might something broken and this returns null.
-  Future<String> setPower(String userID, int power) async {
-    final powerMap = Map<String, Object?>.from(
-      getState(EventTypes.RoomPowerLevels)?.content ?? {},
-    );
+  Future<String> setPower(String userId, int power) async {
```
Comment (td-famedly) [inline lib/src/room.dart:1190]: a test would be nice to have :D

File: lib/src/room.dart [hunk 2]
```diff
@@ -1187,22 +1187,26 @@ class Room {
...
+    final powerLevelMapCopy =
+        getState(EventTypes.RoomPowerLevels)?.content.copy() ?? {};
+
+    final usersPowerMap =
+        powerLevelMapCopy.tryGetMap<String, int>('users') ?? {};
 
-    final usersPowerMap = powerMap['users'] is Map<String, Object?>
-        ? powerMap['users'] as Map<String, Object?>
-        : (powerMap['users'] = <String, Object?>{});
+    if (power == getDefaultPowerLevel(powerLevelMapCopy)) {
```
Comment (td-famedly) [inline lib/src/room.dart:1197]: Couldn't find anything in the spec which suggests we should do this, but probably fine. Can we add a test for this as well? :D

File: lib/src/room.dart (hunk 1 above)
Comment (krille-chan) [reply lib/src/room.dart:1190]: Added a test :)

**Review by td-famedly — APPROVED** [review]
lg

**Review by nico-famedly — APPROVED** [review]
Strictly speaking I guess there could be an edge case if the users map is set to a list, where we won't set the permissions ever now. I think those rooms probably fail to federate and so this is fine and we don't need to care about it :D

File: lib/src/room.dart [hunk 3]
```diff
@@ -1187,22 +1187,18 @@ class Room {
...
-    final powerMap = Map<String, Object?>.from(
-      getState(EventTypes.RoomPowerLevels)?.content ?? {},
-    );
-
-    final usersPowerMap = powerMap['users'] is Map<String, Object?>
-        ? powerMap['users'] as Map<String, Object?>
-        : (powerMap['users'] = <String, Object?>{});
+  Future<String> setPower(String userId, int power) async {
+    final powerLevelMapCopy =
+        getState(EventTypes.RoomPowerLevels)?.content.copy() ?? {};
```
Comment (nico-famedly) [inline lib/src/room.dart:1192]: Can you also apply this fix to `enableGroupCalls()` (and possibly a few others)?

File: lib/src/room.dart [hunk 4]
```diff
@@ -1187,22 +1187,18 @@ class Room {
...
 
-    usersPowerMap[userID] = power;
+    powerLevelMapCopy['users'] ??= <String, Object?>{};
+    powerLevelMapCopy.tryGetMap<String, Object?>('users')?[userId] = power;
```
Comment (nico-famedly) [inline lib/src/room.dart:1195]: This operates on a copy of `powerLevelMapCopy`, so this actually never modifies the power levels of that user! That's probably also why the tests hang, but for me at least the command test with the op command explicitly fails. Maybe we can add a proper test for this?

File: lib/src/room.dart [hunk 5]
```diff
@@ -1187,22 +1187,18 @@ class Room {
...
 
     return await client.setRoomStateWithKey(
```
Comment (nico-famedly) [inline lib/src/room.dart:1197]: this actually now means that the changes will only be reflected in the state event later. I would bet there is some code that depends on this function having immediately reflected the changes. Should we consider this a breaking change or maybe wait for the event to come down in sync? Maybe we can just ignore that, but I think it makes sense to consider the impact of that. For example the following code would now only set the permissions for one of the users:

```dart
await room.setPower('@a:example.org', 10);
await room.setPower('@b:example.org', 10);
```

File: lib/src/room.dart (hunk 3 above)
Comment (krille-chan) [reply lib/src/room.dart:1192]: Would prefer to do this in a different PR 🤔 one thing after another

File: lib/src/room.dart (hunk 4 above)
Comment (krille-chan) [reply lib/src/room.dart:1195]: this is also a copy? 😳 oopsie...

File: lib/src/room.dart (hunk 5 above)
Comment (nico-famedly) [reply lib/src/room.dart:1210]: We discussed this and agreed that just documenting it is the best solution for now.
